src/tools/w3dviewer/w3dviewdoc.cpp

```diff
@@ -0,0 +1,957 @@
+
+// clang-format off
+BEGIN_MESSAGE_MAP(CW3DViewDoc, CDocument)
+END_MESSAGE_MAP()
+// clang-format on
+
+CW3DViewDoc::CW3DViewDoc() :
+    m_scene(nullptr),
+    m_textureScene(nullptr),
+    m_backgroundScene(nullptr),
+    m_model(nullptr),
+    m_backgroundObject(nullptr),
+    m_animation(nullptr),
+    m_light(nullptr),
+    m_backgroundBitmap(nullptr),
+    m_textureCamera(nullptr),
+    m_docCamera(nullptr),
+    m_fogColor(0.5f, 0.5f, 0.5f),
+    m_isQCompressed(false),
+    m_compressType(2),
+    m_frameCount(0.0f),
+    m_blendFrames(TRUE),
+    m_animateCamera(false),
+    m_resetCamera(true),
+    m_forceCameraReset(true),
+    m_useManualFov(false),
+    m_useManualClipPlanes(false),
+    m_initialized(false),
+    m_fogEnabled(false),
+    m_time(0.0f)
+{
+    m_animateCamera = theApp.GetProfileInt("Config", "AnimateCamera", 0) == 1;
+    m_resetCamera = theApp.GetProfileInt("Config", "ResetCamera", 1) == 1;
+}
+
+CW3DViewDoc::~CW3DViewDoc()
+{
+    Free();
+}
+
+BOOL CW3DViewDoc::OnNewDocument()
+{
+    BOOL result = CDocument::OnNewDocument();
+
+    if (result) {
+        m_loadedFiles.Delete_All();
+        m_forceCameraReset = true;
+
+        if (m_scene != nullptr && m_model != nullptr) {
+            RemoveRenderObject(m_model);
+        }
+
+        if (m_scene != nullptr) {
+            m_scene->Remove_All_LOD_Objects();
+            m_scene->Set_Fog_Color(m_fogColor);
+        }
+
+        if (m_model != nullptr) {
+            Ref_Ptr_Release(m_animation);
+            Ref_Ptr_Release(m_model);
+        }
+
+        CMainFrame *frame = reinterpret_cast<CMainFrame *>(AfxGetMainWnd());
+
+        if (frame != nullptr) {
+            CDataTreeView *view = reinterpret_cast<CDataTreeView *>(frame->m_splitter.GetPane(0, 0));
+
+            if (view != nullptr) {
+                view->GetTreeCtrl().DeleteItem(TVI_ROOT);
+                view->AddCategories();
+            }
+        }
+
+        Free();
+        W3DAssetManager::Get_Instance()->Free_Assets();
+
+        return TRUE;
+    }
+
+    return result;
+}
+
+BOOL CW3DViewDoc::OnOpenDocument(LPCTSTR lpszPathName)
+{
+    BOOL result = CDocument::OnOpenDocument(lpszPathName);
+    if (result) {
+        CGraphicView *view = GetCurrentGraphicView();
+
+        if (view != nullptr) {
+            view->EnableRendering(false);
+        }
+
+        LoadFile(lpszPathName);
+
+        CMainFrame *frame = reinterpret_cast<CMainFrame *>(AfxGetMainWnd());
+
+        if (frame != nullptr) {
+            CDataTreeView *tree = reinterpret_cast<CDataTreeView *>(frame->m_splitter.GetPane(0, 0));
+
+            if (tree != nullptr) {
+                tree->AddRenderObjects();
+            }
+        }
+
+        if (view != nullptr) {
+            view->EnableRendering(true);
+        }
+
+        return TRUE;
+    }
+
+    return result;
+}
+
+void CW3DViewDoc::Free()
+{
+    if (m_textureScene != nullptr) {
+        if (m_backgroundBitmap != nullptr) {

```

Comment:
Where is this allocated? Unable to find.

Reply:
Its allocated by code I haven't cloned yet.